### ISSUE TITLE
Preview stuck on the loading screen 

### ISSUE THREAD

**Issue by drewlonious** [prismicio/prismic-gatsby#282]
Followed the directions, but can't for the life of me get previews to work.
The preview is stuck on "loading" and the console says the following, which is crazy cause Export Event and Page is definitely on those templates. Is this only me?

`./src/pages/404.js 61:12-17"export 'Event' was not found in '../templates/event.js'./src/pages/404.js 60:10-14"export 'Page' was not found in '../templates/page.js'./src/components/Header.js`

my 404.js

```
import React from "react"
import Helmet from 'react-helmet'
import { withUnpublishedPreview } from 'gatsby-source-prismic'
import {Page} from '../templates/page'
import {Event} from '../templates/event'

// import Layout from "../components/layout"
import SEO from "../components/seo"

const NotFoundPage = () => (
  <>
    <Helmet
      bodyAttributes={{
          class: 'page page-404'
      }}
    />
    <SEO title="404: Not found" />
    <h1>NOT FOUND</h1>
    <p>You just hit a route that doesn&#39;t exist... the sadness.</p>
  </>
)

// If an unpublished `page` document is previewed, PageTemplate will be rendered.
export default withUnpublishedPreview(NotFoundPage, {
  templateMap: {
    page: Page,
    events: Event,
  },
})

```

my templates/page.js
```
import React from "react"
import { graphql } from "gatsby"
import Helmet from "react-helmet"
import ImageHero from "../components/ImageHero"
import Card from "../components/Card"
// import { SiteInfo, Events } from "../components/slices"
import SliceZone from "../components/SliceZone"
import { withPreview } from 'gatsby-source-prismic'


const Page = ({ data: { page, events, faqs, allpages } }) => {
 
  return (
    <>
      <Helmet
        bodyAttributes={{
          class: "post event-post",
        }}
      />

      <ImageHero
        image={require("../assets/wallpaper.jpg")}
        video={require("../assets/kk-vid.mp4")}
        title={page.data.page_title.text}
        text={page.data.page_description.text}
      />

      <SliceZone slices={page.data.body} />

    </>
  )
}

export default withPreview(Page)

export const pageQuery = graphql`
  query PageBySlug($uid: String!) {
    page: prismicPage(uid: { eq: $uid }) {
      id
      uid
      prismicId
      data {
        page_description {
          text
        }
        page_title {
          text
        }
        body {
          ... on PrismicPageBodySiteInfo {
            id
            slice_type
            primary {
              text {
                text
              }
            }
          }
          ... on PrismicPageBodyEvents {
            id
            slice_type
            primary {
              title {
                text
              }
            }
          }
        }
      }
    }
    allpages: allPrismicPage {
      nodes {
        data {
          body {
            ... on PrismicPageBodySiteInfo {
              id
              slice_type
              primary {
                image {
                  url
                }
                text {
                  text
                }
              }
            }
            ... on PrismicPageBodyEvents {
              id
              slice_type
              primary {
                title {
                  text
                }
              }
            }
          }
        }
      }
      edges {
        node {
          data {
            body {
              __typename
              ... on PrismicPageBodySiteInfo {
                id
                slice_type
                primary {
                  text {
                    text
                  }
                  image {
                    url
                  }
                }
              }
              ... on PrismicPageBodyEvents {
                id
                slice_type
                primary {
                  title {
                    text
                  }
                }
              }
            }
          }
        }
      }
    }
    events: allPrismicEvent(
      sort: { fields: [first_publication_date], order: DESC }
    ) {
      nodes {
        data {
          title {
            text
          }
          text {
            text
          }
          featured_image {
            url
          }
          date(formatString: "MM-DD-YY")
        }
        url
        uid
        id
        first_publication_date
      }
    }
    faqs: prismicFaq {
      data {
        group {
          answer {
            text
          }
          question {
            text
          }
        }
      }
    }
  }
`

```

linkresolver.js
```
const linkResolver = ({ node, key, value }) => doc => {
  if (doc.type === "event") {
    return `/events/${doc.uid}`
  }
  if (doc.type === "page") {
    return `/${doc.uid}`
  }
  if (doc.type === "team_members") {
    return `/team/${doc.uid}`
  }

  return `/${doc.uid}`
}

module.exports = linkResolver
```

**Comment by jackkemmish** [comment]
Have you tried removing the brackets around `Page` and `Event`?

But I am also having the issue of being stuck on the loading page. I have followed the Preview HOC set up but am not seeing a preview on the load page.

Any help would be greatly appreciated!

**Comment by bndnio** [comment]
For those coming across this now, I just ran into this issue and worked it out 🙂  

What's going on is that the `linkResolver` as documented has a different signature than that which is expected by `withPreview` (it expects the `func(obj) -> func(obj) -> str` signature, similar to the one [set in `gatsby-config.js`](https://github.com/angeloashmore/gatsby-source-prismic#how-to-use), rather than the `func(obj) -> str` signature exported by the [`linkResolver.js` function in the prismic setup docs](https://prismic.io/docs/technologies/link-resolver-javascript))

Long story short, replace the code found in [the preview setup docs](https://github.com/angeloashmore/gatsby-source-prismic/blob/master/docs/previews.md#withpreviewresolver)
```js
export default withPreviewResolver(PreviewPage, {
  repositoryName: "your-prismic-repository-name",
  linkResolver,
})
```
with
```js
export default withPreviewResolver(PreviewPage, {
  repositoryName: "your-prismic-repository-name",
  linkResolver: () => linkResolver,
})
```

**Comment by bndnio** [comment]
A follow-up gotcha that had me, is the need to modify how I'm querying data in order for the preview data merge to work. Based on the recommendation of the previous prismic gatsby source plugin I used (gatsby-source-prismic-graphql), I had to change

```graphql
allPrismicXPage {
  edges {
    node {
      ...
      data {
        ...
      }
    }
  }
}
```
to 
```graphql
prismicXPage {
  ...
  data {
    ...
  }
}
```

Now that I recognize it, this new syntax is far nicer anyway 😅

**Comment by bndnio** [comment]
@angeloashmore I'm happy to open a PR to fix the docs, but I'm unsure which way is intended. Do you prefer that:
1. `linkResolver` has the signature `func(obj) -> func(obj) -> str` and the `gatsby-config.js` setup doc should change, or
2. `linkResolver` has the signature `func(obj) -> str` and the `preview` docs should change

I would be inclined to lean towards 2 because that is how [prismic docs portray the `linkResolver`](https://prismic.io/docs/technologies/link-resolver-gatsby#1.-create-the-link-resolver), but I don't totally understand the use of the function when curried again and called with `({ node, key, value })` -- perhaps you could shed some light on it?

**Comment by angeloashmore** [comment]
Hey @bndnio, thanks for the offer to update the docs!

The link resolver function needs to have the following signature for both `gatsby-config.js` and the preview HOCs:

```
({ node, key, value }) => (doc) => string
```

The `{ node, key, value }` portion is specific to this plugin and is intended to provide additional data that normally isn't available when using Prismic outside of Gatsby. This would allow you to use a different link resolver for specific fields as necessary.

In the next major version, V4, the `{ node, key, value }` data will be removed from link resolvers as it is a common source of confusion. You can read a little bit more about the change here: [gatsby-source-prismic v4 Roadmap](https://labs.walltowall.com/gatsby-source-prismic-v4-roadmap/#simpler-link-resolver-and-html-serializer-signatures).

You're right that the Prismic docs for Gatsby show the incorrect link resolver signature in their `/src/utils/linkResolver.js` example. It does correctly provide it in their `gatsby-config.js`.

Until V4 is out, perhaps the best solution is to add a note to the [`withPreviewResolver`](https://github.com/angeloashmore/gatsby-source-prismic/blob/c86470cf30a51dfd8607106d15e93f2521b87d80/docs/previews.md#withpreviewresolver) section of the docs. Something like:

> The `linkResolver` function must be the same as the one provided in your `gatsby-config.js` file. This means the `linkResolver` function should look something like:
> ```
> ({ node, key, value }) => (doc) => `/${doc.id}`
> ```

Feel free to expand on that as you see fit. Thanks!

**Comment by angeloashmore** [comment]
I'm cleaning out old issues and closing this due to its age. If this is still a problem, please open a new issue and I will take a look!
